README.rst
===============
django-jsonview
===============


.. image:: https://travis-ci.org/jsocol/django-jsonview.png?branch=master
   :target: https://travis-ci.org/jsocol/django-jsonview

**django-jsonview** is a simple decorator that translates Python objects
to JSON and makes sure your view will always return JSON.

I've copied and pasted this so often I decided I just wanted to put it
in a package.


Installation
============

Just install with ``pip``::

    pip install django-jsonview

No need to add to ``INSTALLED_APPS`` or anything.


Usage
=====

Just import the decorator, use, and return a JSON-serializable object::

    from jsonview.decorators import json_view

    @json_view
    def my_view(request):
        return {
            'foo': 'bar',
        }


`Class-based views`_ (CBVs) can inherit from JsonView, use Django's
``@method_decorator`` or wrap the output of ``.as_view()``::

    # inherit from JsonView
    from jsonview.views import JsonView


    class MyView(JsonView):
        def get_context_data(self, **kwargs):
            context = super(MyView, self).get_context_data(**kwargs)
            context['my_key'] = 'some value'
            return context

    # or, method decorator
    from django.utils.decorators import method_decorator
    from jsonview.decorators import json_view


    class MyView(View):
        @method_decorator(json_view)
        def dispatch(self, *args, **kwargs):
            return super(MyView, self).dispatch(*args, **kwargs)

    # or, in URLconf

    patterns = [
        url(r'^/my-view/$', json_view(MyView.as_view())),
    ]


Content Types
-------------

If you need to return a content type other than the standard
``application/json``, you can specify that in the decorator with the
``content_type`` argument, for example::

    from jsonview.decorators import json_view

    @json_view(content_type='application/vnd.github+json')
    def myview(request):
        return {'foo': 'bar'}

The response will have the appropriate content type header.


Return Values
-------------

The default case is to serialize your return value and respond with HTTP
200 and a Content-Type of ``application/json``.

The ``@json_view`` decorator will handle many exceptions and other
cases, including:

* ``Http404``
* ``PermissionDenied``
* ``HttpResponseNotAllowed`` (e.g. ``require_GET``, ``require_POST``)
* ``jsonview.exceptions.BadRequest`` (see below)
* Any other exception (logged to ``django.request``).

Any of these exceptions will return the correct status code (i.e., 404,
403, 405, 400, 500) a Content-Type of ``application/json``, and a
response body that looks like::

    json.dumps({
        'error': STATUS_CODE,
        'message': str(exception),
    })

.. note::

   As of v0.4, application exceptions do **not** behave this way if
   ``DEBUG = False``. When ``DEBUG = False``, the ``message`` value is
   always ``An error occurred``. When ``DEBUG = True``, the exception
   message is sent back.


``BadRequest``
--------------

HTTP does not have a great status code for "you submitted a form that
didn't validate," and so Django doesn't support it very well. Most
examples just return 200 OK.

Normally, this is fine. But if you're submitting a form via Ajax, it's
nice to have a distinct status for "OK" and "Nope." The HTTP 400 Bad
Request response is the fallback for issues with a request
not-otherwise-specified, so let's do that.

To cause ``@json_view`` to return a 400, just raise a
``jsonview.exceptions.BadRequest`` with whatever appropriate error
message.


Exceptions
----------

If your view raises an exception, ``@json_view`` will catch the
exception, log it to the normal ``django.request`` logger_, and return a
JSON response with a status of 500 and a body that looks like the
exceptions in the `Return Values`_ section.

.. note::

   Because the ``@json_view`` decorator handles the exception instead of
   propagating it, any exception middleware will **not** be called, and
   any response middleware **will** be called.


Status Codes
------------

If you need to return a different HTTP status code, just return two
values instead of one. The first is your serializable object, the second
is the integer status code::

    @json_view
    def myview(request):
        if not request.user.is_subscribed():
            # Send a 402 Payment Required status.
            return {'subscribed': False}, 402
        # Send a 200 OK.
        return {'subscribed': True}


Extra Headers
-------------

You can add custom headers to the response by returning a tuple of three
values: an object, a status code, and a dictionary of headers.

::

    @json_view
    def myview(request):
        return {}, 200, {'X-Server': 'myserver'}

Custom header values may be overwritten by response middleware.


Raw Return Values
-----------------

To make it possible to cache JSON responses as strings (and because they
aren't JSON serializable anyway) if you return an ``HttpResponse``
object (or subclass) it will be passed through unchanged, e.g.::

    from django import http
    from jsonview.decorators import JSON

    @json_view
    def caching_view(request):
        kached = cache.get('cache-key')
        if kached:
            return http.HttpResponse(kached, content_type=JSON)
        # Assuming something else populates this cache.
        return {'complicated': 'object'}

.. note::

   ``@require_POST`` and the other HTTP method decorators  work by
   *returning* a response, rather than *raising*, an exception, so
   ``HttpResponseNotAllowed`` is handled specially.


Alternative JSON Implementations
================================

There is a healthy collection of JSON parsing and generating libraries
out there. By default, it will use the old standby, the stdlib ``json``
module. But, if you'd rather use ujson_, or cjson_ or yajl_, you should
go for it. Just add this to your Django settings::

    JSON_MODULE = 'ujson'

Anything, as long as it's a module that has ``.loads()`` and ``.dumps()``
methods.


Configuring JSON Output
-----------------------

Additional keyword arguments can be passed to ``json.dumps()`` via the
``JSON_OPTIONS = {}`` Django setting. For example, to pretty-print JSON
output::

    JSON_OPTIONS = {
        'indent': 4,
    }

Or to compactify it::

    JSON_OPTIONS = {
        'separators': (',', ':'),
    }

jsonview uses ``DjangoJSONEncoder`` by default. To use a different JSON
encoder, use the ``cls`` option::

    JSON_OPTIONS = {
        'cls': 'path.to.MyJSONEncoder',
    }

``JSON_OPTIONS['cls']`` may be a dotted string or a ``JSONEncoder``
class.

**If you are using a JSON module that does not support the ``cls``
kwarg**, such as ujson, set the ``cls`` option to ``None``::

    JSON_OPTIONS = {
        'cls': None,
    }

Default value of content-type is 'application/json'. You can change it
via the ``JSON_DEFAULT_CONTENT_TYPE`` Django settings. For example, to
add charset::

   JSON_DEFAULT_CONTENT_TYPE = 'application/json; charset=utf-8'


Atomic Requests
===============

Because ``@json_view`` catches exceptions, the normal Django setting
``ATOMIC_REQUESTS`` does not correctly cause a rollback. This can be
worked around by explicitly setting ``@transaction.atomic`` *below* the
``@json_view`` decorator, e.g.::

    @json_view
    @transaction.atomic
    def my_func(request):
        # ...


Contributing
============

`Pull requests`_ and issues_ welcome! I ask two simple things:

- Tests, including the new ones you added, must pass. (See below.)
- Coverage should not drop below 100. You can install ``coverage`` with
  pip and run ``./run.sh coverage`` to check.
- The ``flake8`` tool should not return any issues.


Running Tests
-------------

To run the tests, you probably want to create a virtualenv_, then
install Django and Mock with ``pip``::

    pip install Django==${DJANGO_VERSION} mock==1.0.1

Then run the tests with::

    ./run.sh test


.. _logger:
   https://docs.djangoproject.com/en/dev/topics/logging/#django-request
.. _Pull requests: https://github.com/jsocol/django-jsonview/pulls
.. _issues: https://github.com/jsocol/django-jsonview/issues
.. _virtualenv: http://www.virtualenv.org/
.. _ujson: https://pypi.python.org/pypi/ujson
.. _cjson: https://pypi.python.org/pypi/python-cjson
.. _yajl: https://pypi.python.org/pypi/yajl
.. _Class-based views: https://docs.djangoproject.com/en/1.9/topics/class-based-views/intro/#decorating-class-based-views


CHANGELOG
Changelog
=========

v2.0.0
------

- Update Django versions to 2.1, 2.2, 3.0

v1.3.1
------

- Remove Sphinx-specific directives from README


v1.3.0
------

- Propagate exception details to signal (#43)
- Test against Django 2.1 and 2.2

v1.2.0
------

- Add JsonView CBV base class (#39)

v1.1.0
------

- Add `JSON_DEFAULT_CONTENT_TYPE` setting
- Update compatibility/test for Django 1.8, 1.10, 1.11.
- Fix issue #36 mutating settings.


v1.0.0
------

- Remove deprecated JSON_USE_DJANGO_SERIALIZER setting.


v0.5.1
------

- Drop unnecesary log message on module init.
- Drop compatibility for Django<1.8.


v0.5.0
------

- Deprecate JSON_USE_DJANGO_SERIALIZER in favor of new JSON_OPTIONS
  setting.
- Improve exception handling and compatibility with mail_admins.
- Include traceback in response when settings.DEBUG is True.
- Improve Django 1.8 and Python 3 compatibility.


v0.4.3
------

- Add setting to make Django's serializer optional when using 3rd party
  JSON modules.


v0.4.2
------

- Add support for datetime handling.


v0.4.1
------

- Fix trove classifiers for Python 3.


v0.4.0
------

- Supports new Djangos: 1.4.x, 1.5.x, 1.6.x, and 1.7bx.
- Accepts a content_type kwarg to change from default application/json.
- Pass HttpResponse objects through verbatim. (Previously choked.)
- Only send exception text when settings.DEBUG is True.


v0.3.0
------

- Python3 support


v0.2.2
------

- Can now specify a json module instead of always using standard
  library.


v0.2.1
------

- Fix issue handling unicode in exceptions.


v0.2.0
------

- Propertly emit exception signal to play nicely with e.g. Raven.
- Can return custom headers along with response code.
- Follow more Django core conventions for logging.
- Testing infrasctructure clean up.


v0.1.1
------

- First release


test_settings.py
import logging


logging.disable(logging.CRITICAL)

INSTALLED_APPS = ('jsonview',)

SECRET_KEY = 'foo'

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'test.db',
    },
}


jsonview/tests.py
from __future__ import absolute_import, unicode_literals

import json
import sys
from unittest import SkipTest

import django
from django import http
from django.core.exceptions import PermissionDenied
from django.core.serializers.json import DjangoJSONEncoder
from django.test import RequestFactory, TestCase
from django.test.utils import override_settings
from django.utils import timezone
from django.utils.decorators import method_decorator
from django.views.decorators.http import require_POST
from django.views.generic import View

import mock

from .decorators import json_view, json_request
from .exceptions import BadRequest
from .views import JsonView


JSON = 'application/json'
rf = RequestFactory()


def eq_(a, b, msg=None):
    """From nose.tools.eq_."""
    assert a == b, msg or '%r != %r' % (a, b)


if sys.version < '3':
    def b(x):
        return x
else:
    import codecs

    def b(x):
        return codecs.latin_1_encode(x)[0]


class CustomTestEncoder(DjangoJSONEncoder):
    def default(self, o):
        try:
            return o.for_json()
        except AttributeError:
            return super(CustomTestEncoder, self).default(o)


class JsonViewTests(TestCase):
    def test_object(self):
        data = {
            'foo': 'bar',
            'baz': 'qux',
            'quz': [{'foo': 'bar'}],
        }

        @json_view
        def temp(req):
            return data

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        eq_(data, json.loads(res.content.decode('utf-8')))
        eq_(JSON, res['content-type'])

    def test_list(self):
        data = ['foo', 'bar', 'baz']

        @json_view
        def temp(req):
            return data

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        eq_(data, json.loads(res.content.decode('utf-8')))
        eq_(JSON, res['content-type'])

    def test_404(self):
        @json_view
        def temp(req):
            raise http.Http404('foo')

        res = temp(rf.get('/'))
        eq_(404, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_(404, data['error'])
        eq_('foo', data['message'])

    def test_permission(self):
        @json_view
        def temp(req):
            raise PermissionDenied('bar')

        res = temp(rf.get('/'))
        eq_(403, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_(403, data['error'])
        eq_('bar', data['message'])

    def test_bad_request(self):
        @json_view
        def temp(req):
            raise BadRequest('baz')

        res = temp(rf.get('/'))
        eq_(400, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_(400, data['error'])
        eq_('baz', data['message'])

    def test_not_allowed(self):
        @json_view
        @require_POST
        def temp(req):
            return {}

        res = temp(rf.get('/'))
        eq_(405, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_(405, data['error'])

        res = temp(rf.post('/'))
        eq_(200, res.status_code)

    @override_settings(DEBUG=True)
    def test_server_error_debug(self):
        @json_view
        def temp(req):
            raise TypeError('fail')

        res = temp(rf.get('/'))
        eq_(500, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_(500, data['error'])
        eq_('fail', data['message'])
        assert 'traceback' in data

    @override_settings(DEBUG=False)
    def test_server_error_no_debug(self):
        @json_view
        def temp(req):
            raise TypeError('fail')

        res = temp(rf.get('/'))
        eq_(500, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_(500, data['error'])
        eq_('An error occurred', data['message'])

    def test_http_status(self):
        @json_view
        def temp(req):
            return {}, 402
        res = temp(rf.get('/'))
        eq_(402, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_({}, data)

    def test_headers(self):
        @json_view
        def temp(req):
            return {}, 302, {'X-Foo': 'Bar'}
        res = temp(rf.get('/'))
        eq_(302, res.status_code)
        eq_(JSON, res['content-type'])
        eq_('Bar', res['X-Foo'])
        data = json.loads(res.content.decode('utf-8'))
        eq_({}, data)

    def test_signal_sent(self):
        from . import decorators

        @json_view
        def temp(req):
            [][0]  # sic.

        with mock.patch.object(decorators, 'got_request_exception') as s:
            res = temp(rf.get('/'))

        assert s.send.called
        eq_(JSON, res['content-type'])

    @override_settings(DEBUG=True)
    def test_signal_sent_with_propagated_exception(self):
        from django.core.signals import got_request_exception

        def assertion_handler(sender, request, **kwargs):
            for key in ['exception', 'exc_data']:
                assert key in kwargs
            for key in ['error', 'message', 'traceback']:
                assert key in kwargs['exc_data']
            assert isinstance(kwargs['exception'], Exception)

        got_request_exception.connect(assertion_handler)

        @json_view
        def temp(req):
            1/0  # sic.

        temp(rf.get('/'))

    def test_unicode_error(self):
        @json_view
        def temp(req):
            raise http.Http404('page \xe7\xe9 not found')

        res = temp(rf.get('/\xe7\xe9'))
        eq_(404, res.status_code)
        data = json.loads(res.content.decode('utf-8'))
        assert '\xe7\xe9' in data['message']

    def test_override_content_type(self):
        testtype = 'application/vnd.helloworld+json'
        data = {'foo': 'bar'}

        @json_view(content_type=testtype)
        def temp(req):
            return data

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        eq_(data, json.loads(res.content.decode('utf-8')))
        eq_(testtype, res['content-type'])

    def test_passthrough_response(self):
        """Allow HttpResponse objects through untouched."""
        payload = json.dumps({'foo': 'bar'}).encode('utf-8')

        @json_view
        def temp(req):
            return http.HttpResponse(payload, content_type='text/plain')

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        eq_(payload, res.content)
        eq_('text/plain', res['content-type'])

    def test_datetime(self):
        now = timezone.now()

        @json_view
        def temp(req):
            return {'datetime': now}

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        payload = json.dumps({'datetime': now}, cls=DjangoJSONEncoder)
        eq_(b(payload), res.content)

    @override_settings(JSON_OPTIONS={'cls': None})
    def test_datetime_no_serializer(self):
        now = timezone.now()

        @json_view
        def temp(req):
            return {'datetime': now}

        res = temp(rf.get('/'))
        eq_(500, res.status_code)
        payload = json.loads(res.content.decode('utf-8'))
        eq_(500, payload['error'])

    @override_settings(JSON_OPTIONS={'cls': None})
    def test_dont_mutate_json_settings(self):
        """Don't mutate JSON settings during a request.

        The second request should use the same settings as the first, so we
        should get two 500s in a row.
        """
        now = timezone.now()

        @json_view
        def temp(req):
            return {'datetime': now}

        res = temp(rf.get('/'))
        eq_(500, res.status_code)

        # calling this a second time should still generate a 500 and not
        # use the `DjangoJSONEncoder`
        res2 = temp(rf.get('/'))
        eq_(500, res2.status_code)

    @override_settings(
        JSON_OPTIONS={'cls': 'jsonview.tests.CustomTestEncoder'})
    def test_json_custom_serializer_string(self):
        payload = json.dumps({'foo': 'Custom JSON'}).encode('utf-8')

        class Obj(object):
            def for_json(self):
                return 'Custom JSON'

        @json_view
        def temp(req):
            return {'foo': Obj()}

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        eq_(payload, res.content)

    @override_settings(
        JSON_OPTIONS={'cls': CustomTestEncoder})
    def test_json_custom_serializer_class(self):
        payload = json.dumps({'foo': 'Custom JSON'}).encode('utf-8')

        class Obj(object):
            def for_json(self):
                return 'Custom JSON'

        @json_view
        def temp(req):
            return {'foo': Obj()}

        res = temp(rf.get('/'))
        eq_(200, res.status_code)
        eq_(payload, res.content)

    def test_method_decorator_on_dispatch(self):
        class TV(View):
            @method_decorator(json_view)
            def dispatch(self, *a, **kw):
                return super(TV, self).dispatch(*a, **kw)

            def get(self, request):
                return {'foo': 'bar'}

        view = TV.as_view()
        res = view(rf.get('/'))
        eq_(200, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_('bar', data['foo'])

    def test_method_decorator_on_class(self):
        if django.VERSION < (1, 9):
            raise SkipTest('Feature added in Django 1.9')

        @method_decorator(json_view, name='dispatch')
        class TV(View):
            def get(self, request):
                return {'foo': 'bar'}

        view = TV.as_view()
        res = view(rf.get('/'))
        eq_(200, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_('bar', data['foo'])

    def test_view_class_get(self):
        class MyView(JsonView):
            def get_context_data(self, **kwargs):
                context = super(MyView, self).get_context_data(**kwargs)
                context['foo'] = 'bar'
                return context

        view = MyView.as_view()

        res = view(rf.get('/'))
        eq_(200, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_('bar', data['foo'])

    def test_view_class_post(self):
        class MyView(JsonView):
            def post(self, request, *args, **kwargs):
                return self.get(request, *args, **kwargs)

        view = MyView.as_view()

        res = view(rf.post('/'))
        eq_(200, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_({}, data)

    def test_wrap_as_view(self):
        class TV(View):
            def get(self, request):
                return {'foo': 'bar'}

        view = json_view(TV.as_view())
        res = view(rf.get('/'))
        eq_(200, res.status_code)
        eq_(JSON, res['content-type'])
        data = json.loads(res.content.decode('utf-8'))
        eq_('bar', data['foo'])


class JsonRequestTests(TestCase):
    def test_application_json(self):
        data = {
            'foo': 'bar',
            'baz': 'qux',
            'quz': [{'foo': 'bar'}],
        }

        @json_request
        def temp(req):
            return req.data

        res = temp(rf.post(
                '/',
                data=json.dumps(data),
                content_type='application/json'
            ))
        eq_(res, data)

    def test_get_requests(self):
        data = {
            'foo': 'bar',
            'baz': '0'
        }

        @json_request(assume_json=False)
        def temp(req):
            return req.data

        res = temp(rf.get('/?foo=bar&baz=0'))
        eq_(res, data)

    def test_post_requests(self):
        data = {
            'foo': 'bar',
            'baz': '0'
        }

        @json_request(assume_json=False)
        def temp(req):
            return req.data

        # test application/x-www-form-urlencoded
        res = temp(rf.post(
                '/',
                data='foo=bar&baz=0',
                content_type='application/x-www-form-urlencoded'
            ))
        eq_(res, data)

        # test multipart/form-data
        res = temp(rf.post('/', data=data, files=None))
        eq_(res, data)

    def test_assume_json(self):
        data = {
            'foo': 'bar',
            'baz': '0'
        }

        @json_request(assume_json=True)
        def temp(req):
            return req.data

        @json_request(assume_json=False)
        def temp_2(req):
            return req.data

        # test get request
        res = temp(rf.get('/?foo=bar&baz=0'))
        eq_(res, {})

        res = temp_2(rf.get('/?foo=bar&baz=0'))
        eq_(res, data)

        # test application/x-www-form-urlencoded
        res = temp(rf.post(
                '/',
                data='foo=bar&baz=0',
                content_type='application/x-www-form-urlencoded'
            ))
        eq_(res, {})

        res = temp_2(rf.post(
                '/',
                data='foo=bar&baz=0',
                content_type='application/x-www-form-urlencoded'
            ))
        eq_(res, data)

        # test multipart/form-data
        res = temp(rf.post('/', data=data, files=None))
        eq_(res, {})

        res = temp_2(rf.post('/', data=data, files=None))
        eq_(res, data)


run.sh
#!/bin/bash

export PYTHONPATH=".:$PYTHONPATH"
export DJANGO_SETTINGS_MODULE="test_settings"

PROG="$0"
CMD="$1"
shift

usage() {
    echo "USAGE: $PROG [command]"
    echo "  test - run the jsonview tests"
    echo "  shell - open the Django shell"
    echo "  check - run flake8"
    echo "  coverage - run tests with coverage"
    exit 1
}

case "$CMD" in
    "test" )
        django-admin.py test jsonview $@
        ;;
    "shell" )
        django-admin.py shell $@
        ;;
    "check" )
        flake8 jsonview $@
        ;;
    "coverage" )
        coverage run `which django-admin.py` test jsonview $@
        coverage report \
            -m \
            --include=jsonview/* \
            --omit=jsonview/tests.py \
            --fail-under=100
        ;;
    * )
        usage ;;
esac


jsonview/compat.py
from __future__ import absolute_import, unicode_literals

import sys
import types

try:
    from importlib import import_module
except ImportError:
    from django.utils.importlib import import_module

try:
    from django.utils.module_loading import import_string
except ImportError:
    def import_string(dotted_path):
        try:
            module_path, obj_name = dotted_path.rsplit('.', 1)
        except ValueError:
            msg = '%s does not look like a module path' % dotted_path
            raise ImportError(msg)

        module = import_module(module_path)

        try:
            return getattr(module, obj_name)
        except AttributeError:
            msg = 'Module %s does not define name %s' % (module_path, obj_name)
            raise ImportError(msg)


PY3 = sys.version_info[0] == 3

if PY3:
    binary_type = bytes
    class_types = (type,)
    string_types = (str,)
    text_type = str
else:
    binary_type = str
    class_types = (type, types.ClassType)
    string_types = (basestring,)  # noqa
    text_type = unicode  # noqa


jsonview/decorators.py
from __future__ import absolute_import, unicode_literals

import logging
import traceback
from functools import wraps
from importlib import import_module

from django import http
from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.core.handlers.base import BaseHandler
from django.core.serializers.json import DjangoJSONEncoder
from django.core.signals import got_request_exception
import six
from django.utils.module_loading import import_string

from .exceptions import BadRequest

json = import_module(getattr(settings, 'JSON_MODULE', 'json'))
JSON = getattr(settings, 'JSON_DEFAULT_CONTENT_TYPE', 'application/json')
logger = logging.getLogger('django.request')


def _dump_json(data):
    options = {}
    options.update(getattr(settings, 'JSON_OPTIONS', {}))

    # Use the DjangoJSONEncoder by default, unless cls is set to None.
    options.setdefault('cls', DjangoJSONEncoder)
    if isinstance(options['cls'], six.string_types):
        options['cls'] = import_string(options['cls'])
    elif options['cls'] is None:
        options.pop('cls')

    return json.dumps(data, **options)


def json_view(*args, **kwargs):
    """Ensure the response content is well-formed JSON.

    Views wrapped in @json_view can return JSON-serializable Python objects,
    like lists and dicts, and the decorator will serialize the output and set
    the correct Content-type.

    Views may also throw known exceptions, like Http404, PermissionDenied, etc,
    and @json_view will convert the response to a standard JSON error format,
    and set the status code and content type.

    If you return a two item tuple, the first is a JSON-serializable object and
    the second is an integer used for the HTTP status code, e.g.:

    >>> @json_view
    ... def example(request):
    ...    return {'foo': 'bar'}, 418

    By default all responses will get application/json as their content type.
    You can override it for non-error responses by giving the content_type
    keyword parameter to the decorator, e.g.:

    >>> @json_view(content_type='application/vnd.example-v1.0+json')
    ... def example2(request):
    ...     return {'foo': 'bar'}

    """

    content_type = kwargs.get('content_type', JSON)

    def decorator(f):
        @wraps(f)
        def _wrapped(request, *a, **kw):
            try:
                status = 200
                headers = {}
                ret = f(request, *a, **kw)

                if isinstance(ret, tuple):
                    if len(ret) == 3:
                        ret, status, headers = ret
                    else:
                        ret, status = ret

                # Some errors are not exceptions. :\
                if isinstance(ret, http.HttpResponseNotAllowed):
                    blob = _dump_json({
                        'error': 405,
                        'message': 'HTTP method not allowed.'
                    })
                    return http.HttpResponse(
                        blob, status=405, content_type=JSON)

                # Allow HttpResponses to go straight through.
                if isinstance(ret, http.HttpResponse):
                    return ret

                blob = _dump_json(ret)
                response = http.HttpResponse(blob, status=status,
                                             content_type=content_type)
                for k in headers:
                    response[k] = headers[k]
                return response
            except http.Http404 as e:
                blob = _dump_json({
                    'error': 404,
                    'message': six.text_type(e),
                })
                logger.warning('Not found: %s', request.path,
                               extra={
                                   'status_code': 404,
                                   'request': request,
                               })
                return http.HttpResponseNotFound(blob, content_type=JSON)
            except PermissionDenied as e:
                logger.warning(
                    'Forbidden (Permission denied): %s', request.path,
                    extra={
                        'status_code': 403,
                        'request': request,
                    })
                blob = _dump_json({
                    'error': 403,
                    'message': six.text_type(e),
                })
                return http.HttpResponseForbidden(blob, content_type=JSON)
            except BadRequest as e:
                blob = _dump_json({
                    'error': 400,
                    'message': six.text_type(e),
                })
                return http.HttpResponseBadRequest(blob, content_type=JSON)
            except Exception as e:
                exc_data = {
                    'error': 500,
                    'message': 'An error occurred',
                }
                if settings.DEBUG:
                    exc_data['message'] = six.text_type(e)
                    exc_data['traceback'] = traceback.format_exc()

                blob = _dump_json(exc_data)

                # Generate the usual 500 error email with stack trace and full
                # debugging information
                logger.error(
                    'Internal Server Error: %s', request.path,
                    exc_info=True,
                    extra={
                        'status_code': 500,
                        'request': request
                    }
                )

                # Here we lie a little bit. Because we swallow the exception,
                # the BaseHandler doesn't get to send this signal. It sets the
                # sender argument to self.__class__, in case the BaseHandler
                # is subclassed.
                got_request_exception.send(
                    sender=BaseHandler,
                    request=request,
                    exception=e,
                    exc_data=exc_data
                )
                return http.HttpResponseServerError(blob, content_type=JSON)
        return _wrapped

    if len(args) == 1 and callable(args[0]):
        return decorator(args[0])
    else:
        return decorator
